pilot/pkg/serviceregistry/kube/controller/ambientindex.go

```diff
@@ -86,6 +87,48 @@ func (c *Controller) generatePodUID(p *v1.Pod) string {
 	return c.clusterID.String() + "//" + "Pod/" + p.Namespace + "/" + p.Name
 }
 
+// servicesByNamespacedHostname returns the service with the given namespace and hostname. If
+// the provided namespace is a wildcard ("*"), all services with the hostname are returned.
+func (a *AmbientIndex) servicesByNamespacedHostname(namespace, hostname string) []*model.ServiceInfo {
+	out := make([]*model.ServiceInfo, 0)
+	if services := a.serviceByHostname[hostname]; services != nil {
+		if namespace == "*" {
+			// The query was for all namespaces. Return all services with the hostname.
+			out = append(out, services...)
+		} else {
+			// Return only the service with the matching hostname.
+			for _, svc := range services {
+				if svc.Namespace == namespace {
+					out = append(out, svc)
+					break
+				}
+			}
+		}
+	}
+	return out
+}
+
+// workloadsByNamespacedHostname returns the workload with the given namespace and hostname. If
+// the provided namespace is a wildcard ("*"), all workloads with the hostname are returned.
+func (a *AmbientIndex) workloadsByNamespacedHostname(namespace, hostname string) []*model.WorkloadInfo {
```

Comment:
Hmm... I think workload hostname is globally unique and maybe not need the namespace part. Maybe not though. Is it only headless services?

Reply:
Yeah I wasn't sure if there could be cases where workload hostname could be scoped to a namespace, same as services. I'm happy to remove the namespace component for now to keep it simple.